### ISSUE TITLE
Reading ST19 on Micrologix 1400

### ISSUE THREAD

**Issue by paulnorstrom** [libplctag/libplctag.NET#199]
Hi, 

Im trying to read ST19:5 on a Micrologix 1400
![bild](https://user-images.githubusercontent.com/80525331/137140159-73e50516-8f5d-431a-8249-af51db556ccb.png)

Ideally I would like to read it as string but I cant without modifying the StringPlcMapper to this: 

`        
        
        string MicroLogixDecode(Tag tag, int offset)
        {
            const int STRING_LENGTH_HEADER = 2;

            int apparentStringLength = tag.GetInt16(offset);

            var actualStringLength = Math.Min(apparentStringLength, MAX_CONTROLLOGIX_STRING_LENGTH);

            var asciiEncodedString = new byte[actualStringLength];
            for (int ii = 0; ii < actualStringLength; ii++)
            {
                if(ii % 2 == 0)
                    asciiEncodedString[ii] = tag.GetUInt8(STRING_LENGTH_HEADER + ii + 1);
                else
                    asciiEncodedString[ii] = tag.GetUInt8(STRING_LENGTH_HEADER + ii - 1);
            }

            return Encoding.ASCII.GetString(asciiEncodedString);
        }
`

So basicly reading the apparentStringLength as int16 instead of 32. 
It also looks like the endianess is wrong so I needed to flip the bytes before creating the string. 

So my question is this, should I create my own TypeXPlcMapper and add this code there or is this something that could be added to the library?

**Comment by kyle-github** [comment]
That method should probably be rewritten to use the core library string functions.   Those will let you get the string length independently of the underlying PLC.   PCCC-based PLCs like MicroLogix use strings of 84 bytes.   2 bytes for the length and 82 bytes for the string data.   The data is byte-swapped on MicroLogix IIRC.

Get the string length with plc_tag_get_string_length(), allocate a local buffer with that length, then call plc_tag_get_string() with the local buffer to get the ASCII/utf-8 version of the string.   This is what I do in Java:

```java
    public String getString(int offset) {
        int rc = Tag.PLCTAG_STATUS_OK;

        rc = plc_tag_get_string_length(this.tag_id, offset);

        // did we get a good string?
        if(rc >= 0) {
            byte[] buf = new byte[rc + 1]; // add one for the zero termination

            rc = plc_tag_get_string(this.tag_id, offset, buf, buf.length);

            if(rc == Tag.PLCTAG_STATUS_OK) {
                return Native.toString(buf);
            }
        }

        return null;
    }
```

**Comment by paulnorstrom** [comment]
Just updated to v1.0.12 
I get an exception during initializing though. 

Setting up tag..
Gateway: 172.16.14.90
Path: 1,0
Tag: ST19:5
DataType: string
Protocol: ab_eip
PlcType: MicroLogix
Timeout: 00:00:05
Tag initializing..

Exception: 12

I made a SstringPlcMapper which has the Decode function you mentioned earlier. In version 1.0.12 that worked fine. 

With 1.0.12 I tested with both but I get an exception with both ways.
